File: src-tauri/src/repo/clients.rs
use rusqlite::{params, params_from_iter, types::Value, Connection};

use crate::error::{AppError, AppResult};
use crate::models::{blank_to_none, Client, ClientFilter, ClientInput, Page, CLIENT_COLUMNS};
use crate::query::{self, Conditions};
use crate::util;

const SORTABLE: &[(&str, &str)] = &[
    ("name", "c.full_name"),
    ("code", "c.client_code"),
    ("city", "c.city"),
    ("created", "c.created_at"),
    ("updated", "c.updated_at"),
    ("policies", "total_policies"),
    ("nextExpiry", "next_expiry"),
];

const DERIVED: &str = "(SELECT COUNT(*) FROM policies p WHERE p.client_id = c.id AND p.status = 'active') AS active_policies, \
     (SELECT COUNT(*) FROM policies p WHERE p.client_id = c.id) AS total_policies, \
     (SELECT MIN(p.expiry_date) FROM policies p WHERE p.client_id = c.id \
        AND p.expiry_date >= date('now', 'localtime')) AS next_expiry";

fn build_conditions(filter: &ClientFilter) -> Conditions {
    let mut c = Conditions::new();

    if !filter.include_archived.unwrap_or(false) {
        c.add_raw("c.is_archived = 0");
    }

    if let Some(search) = filter
        .search
        .as_deref()
        .map(str::trim)
        .filter(|s| !s.is_empty())
    {
        match super::fts_query(search) {
            Some(q) => c.add(
                "c.id IN (SELECT rowid FROM clients_fts WHERE clients_fts MATCH ?)",
                Value::Text(q),
            ),
            None => c.add_many(
                "(c.full_name LIKE ? ESCAPE '\\' OR c.phone LIKE ? ESCAPE '\\')".into(),
                {
                    let pattern = query::like_pattern(search);
                    vec![Value::Text(pattern.clone()), Value::Text(pattern)]
                },
            ),
        }
    }

    if let Some(city) = blank_to_none(filter.city.clone()) {
        c.add("c.city = ?", Value::Text(city));
    }
    if let Some(state) = blank_to_none(filter.state.clone()) {
        c.add("c.state = ?", Value::Text(state));
    }
    if let Some(category) = blank_to_none(filter.category.clone()) {
        c.add(
            "EXISTS (SELECT 1 FROM policies p WHERE p.client_id = c.id AND p.category = ?)",
            Value::Text(category),
        );
    }
    if filter.missing_email.unwrap_or(false) {
        c.add_raw("(c.email IS NULL OR c.email = '')");
    }

    c
}

pub fn list(conn: &Connection, filter: &ClientFilter) -> AppResult<Page<Client>> {
    let conditions = build_conditions(filter);
    let where_sql = conditions.where_sql();

    let total = super::count(
        conn,
        &format!("SELECT COUNT(*) FROM clients c{where_sql}"),
        conditions.params(),
    )?;

    let (page, page_size, limit, offset) = query::paginate(filter.page, filter.page_size);
    let order = query::order_by(
        filter.sort.as_deref(),
        filter.descending.unwrap_or(false),
        SORTABLE,
        "c.full_name",
    );

    let sql = format!(
        "SELECT {CLIENT_COLUMNS}, {DERIVED} FROM clients c{where_sql}{order} LIMIT ? OFFSET ?"
    );
    let mut stmt = conn.prepare(&sql)?;
    let rows = stmt
        .query_map(
            params_from_iter(conditions.params_with([limit, offset])),
            Client::from_row,
        )?
        .collect::<rusqlite::Result<Vec<_>>>()?;

    Ok(Page {
        rows,
        total,
        page,
        page_size,
    })
}

pub fn get(conn: &Connection, id: i64) -> AppResult<Client> {
    let sql = format!("SELECT {CLIENT_COLUMNS}, {DERIVED} FROM clients c WHERE c.id = ?");
    conn.query_row(&sql, params![id], Client::from_row)
        .map_err(|e| match e {
            rusqlite::Error::QueryReturnedNoRows => AppError::NotFound("Client"),
            other => other.into(),
        })
}

pub fn next_client_code(conn: &Connection) -> AppResult<String> {
    let next: i64 = conn.query_row(
        "SELECT IFNULL(MAX(CAST(substr(client_code, 4) AS INTEGER)), 0) + 1 \
         FROM clients WHERE client_code GLOB 'CL-[0-9]*'",
        [],
        |row| row.get(0),
    )?;
    Ok(format!("CL-{next:05}"))
}

fn validate(input: &ClientInput) -> AppResult<()> {
    if input.full_name.trim().is_empty() {
        return Err(AppError::validation("Client name is required"));
    }
    if let Some(email) = blank_to_none(input.email.clone()) {
        if !util::looks_like_email(&email) {
            return Err(AppError::validation(format!(
                "\"{email}\" is not a valid email address"
            )));
        }
    }
    if let Some(dob) = blank_to_none(input.date_of_birth.clone()) {
        if util::parse_date(&dob).is_none() {
            return Err(AppError::validation("Date of birth is not a valid date"));
        }
    }
    Ok(())
}

pub fn create(conn: &Connection, input: &ClientInput) -> AppResult<i64> {
    validate(input)?;

    let code = match blank_to_none(input.client_code.clone()) {
        Some(code) => code,
        None => next_client_code(conn)?,
    };

    conn.execute(
        "INSERT INTO clients (client_code, full_name, email, phone, alt_phone, date_of_birth, \
             gender, address_line1, address_line2, city, state, pincode, occupation, pan, gstin, \
             preferred_language, reminders_opted_out, notes) \
         VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13, ?14, ?15, ?16, ?17, ?18)",
        params![
            code,
            util::tidy_name(&input.full_name),
            blank_to_none(input.email.clone()),
            input.phone.as_deref().and_then(util::normalise_phone),
            input.alt_phone.as_deref().and_then(util::normalise_phone),
            blank_to_none(input.date_of_birth.clone()).and_then(|d| util::parse_date(&d)),
            blank_to_none(input.gender.clone()),
            blank_to_none(input.address_line1.clone()),
            blank_to_none(input.address_line2.clone()),
            blank_to_none(input.city.clone()),
            blank_to_none(input.state.clone()),
            blank_to_none(input.pincode.clone()),
            blank_to_none(input.occupation.clone()),
            blank_to_none(input.pan.clone()).map(|p| p.to_uppercase()),
            blank_to_none(input.gstin.clone()).map(|g| g.to_uppercase()),
            input
                .preferred_language
                .clone()
                .unwrap_or_else(|| "en".into()),
            input.reminders_opted_out.unwrap_or(false) as i64,
            blank_to_none(input.notes.clone()),
        ],
    )
    .map_err(map_unique_error)?;

    Ok(conn.last_insert_rowid())
}

pub fn update(conn: &Connection, id: i64, input: &ClientInput) -> AppResult<()> {
    validate(input)?;

    let changed = conn
        .execute(
            "UPDATE clients SET full_name = ?2, email = ?3, phone = ?4, alt_phone = ?5, \
                 date_of_birth = ?6, gender = ?7, address_line1 = ?8, address_line2 = ?9, \
                 city = ?10, state = ?11, pincode = ?12, occupation = ?13, pan = ?14, gstin = ?15, \
                 preferred_language = ?16, reminders_opted_out = ?17, notes = ?18, \
                 client_code = COALESCE(?19, client_code) \
             WHERE id = ?1",
            params![
                id,
                util::tidy_name(&input.full_name),
                blank_to_none(input.email.clone()),
                input.phone.as_deref().and_then(util::normalise_phone),
                input.alt_phone.as_deref().and_then(util::normalise_phone),
                blank_to_none(input.date_of_birth.clone()).and_then(|d| util::parse_date(&d)),
                blank_to_none(input.gender.clone()),
                blank_to_none(input.address_line1.clone()),
                blank_to_none(input.address_line2.clone()),
                blank_to_none(input.city.clone()),
                blank_to_none(input.state.clone()),
                blank_to_none(input.pincode.clone()),
                blank_to_none(input.occupation.clone()),
                blank_to_none(input.pan.clone()).map(|p| p.to_uppercase()),
                blank_to_none(input.gstin.clone()).map(|g| g.to_uppercase()),
                input
                    .preferred_language
                    .clone()
                    .unwrap_or_else(|| "en".into()),
                input.reminders_opted_out.unwrap_or(false) as i64,
                blank_to_none(input.notes.clone()),
                blank_to_none(input.client_code.clone()),
            ],
        )
        .map_err(map_unique_error)?;

    if changed == 0 {
        return Err(AppError::NotFound("Client"));
    }
    Ok(())
}

pub fn set_archived(conn: &Connection, id: i64, archived: bool) -> AppResult<()> {
    let changed = conn.execute(
        "UPDATE clients SET is_archived = ?2 WHERE id = ?1",
        params![id, archived as i64],
    )?;
    if changed == 0 {
        return Err(AppError::NotFound("Client"));
    }
    Ok(())
}

/// Removes the client together with their policies. The UI confirms first; the
/// archive flag is the softer option offered alongside it.
pub fn delete(conn: &Connection, id: i64) -> AppResult<()> {
    let changed = conn.execute("DELETE FROM clients WHERE id = ?1", params![id])?;
    if changed == 0 {
        return Err(AppError::NotFound("Client"));
    }
    Ok(())
}

pub fn distinct_cities(conn: &Connection) -> AppResult<Vec<String>> {
    let mut stmt = conn.prepare(
        "SELECT DISTINCT city FROM clients WHERE city IS NOT NULL AND city <> '' ORDER BY city",
    )?;
    let rows = stmt
        .query_map([], |row| row.get::<_, String>(0))?
        .collect::<rusqlite::Result<Vec<_>>>()?;
    Ok(rows)
}

/// Finds an existing client by code, then email, then name — used by the importer
/// to decide between updating and inserting.
pub fn find_match(
    conn: &Connection,
    code: Option<&str>,
    email: Option<&str>,
    phone: Option<&str>,
    name: &str,
) -> AppResult<Option<i64>> {
    if let Some(code) = code.filter(|c| !c.is_empty()) {
        if let Some(id) = lookup(conn, "SELECT id FROM clients WHERE client_code = ?1", code)? {
            return Ok(Some(id));
        }
    }
    if let Some(email) = email.filter(|e| !e.is_empty()) {
        if let Some(id) = lookup(
            conn,
            "SELECT id FROM clients WHERE lower(email) = lower(?1)",
            email,
        )? {
            return Ok(Some(id));
        }
    }
    if let Some(phone) = phone.filter(|p| !p.is_empty()) {
        if let Some(id) = lookup(conn, "SELECT id FROM clients WHERE phone = ?1", phone)? {
            return Ok(Some(id));
        }
    }
    lookup(
        conn,
        "SELECT id FROM clients WHERE lower(full_name) = lower(?1)",
        name,
    )
}

fn lookup(conn: &Connection, sql: &str, value: &str) -> AppResult<Option<i64>> {
    let mut stmt = conn.prepare(sql)?;
    let mut rows = stmt.query(params![value])?;
    match rows.next()? {
        Some(row) => Ok(Some(row.get(0)?)),
        None => Ok(None),
    }
}

fn map_unique_error(err: rusqlite::Error) -> AppError {
    match &err {
        rusqlite::Error::SqliteFailure(e, Some(msg))
            if e.code == rusqlite::ErrorCode::ConstraintViolation
                && msg.contains("client_code") =>
        {
            AppError::Conflict("That client code is already in use".into())
        }
        _ => err.into(),
    }
}

File: legacy-windows/src/core/repo/clients.ts
/** A port of `src-tauri/src/repo/clients.rs`. */

import type { Conn } from "../db";
import { AppError, describe } from "../errors";
import { Conditions, likePattern, orderBy, paginate, type Bind } from "../query";
import { blankToNull, boolToInt, toModel, toModels } from "../rows";
import type { Client, ClientFilter, ClientInput, Page } from "../types";
import { looksLikeEmail, normalisePhone, parseDate, tidyName } from "../util";
import { immediateIds } from "./relations";
import { count, ftsQuery, isConstraintViolation } from "./shared";

const SORTABLE: Record<string, string> = {
  name: "c.full_name",
  code: "c.client_code",
  city: "c.city",
  created: "c.created_at",
  updated: "c.updated_at",
  policies: "total_policies",
  nextExpiry: "next_expiry",
};

const COLUMNS =
  "c.id, c.client_code, c.full_name, c.email, c.phone, c.alt_phone, " +
  "c.date_of_birth, c.gender, c.address_line1, c.address_line2, c.city, c.state, c.pincode, " +
  "c.occupation, c.pan, c.gstin, c.preferred_language, c.reminders_opted_out, c.notes, " +
  "c.is_archived, c.created_at, c.updated_at";

const DERIVED =
  "(SELECT COUNT(*) FROM policies p WHERE p.client_id = c.id AND p.status = 'active') AS active_policies, " +
  "(SELECT COUNT(*) FROM policies p WHERE p.client_id = c.id) AS total_policies, " +
  "(SELECT MIN(p.expiry_date) FROM policies p WHERE p.client_id = c.id " +
  "   AND p.expiry_date >= date('now', 'localtime')) AS next_expiry, " +
  "(SELECT COUNT(*) FROM client_relations r " +
  "   WHERE r.client_id = c.id OR r.related_client_id = c.id) AS relatives, " +
  "(SELECT CASE WHEN NOT EXISTS (SELECT 1 FROM policies p WHERE p.client_id = c.id) " +
  "              AND EXISTS (SELECT 1 FROM client_relations r WHERE r.related_client_id = c.id) " +
  "             THEN 1 ELSE 0 END) AS is_dependent";

/**
 * A client with no policy of their own who is listed under somebody else — a
 * spouse on a floater, a dependent child. They are clients like any other and the
 * book holds them once, but a list of two thousand names where half are children
 * is not the book an agent works from, so browsing hides them.
 *
 * The dashboard counts through this too: a child with no email address is not a
 * client the agency is failing to reach.
 */
export const IS_DEPENDENT =
  "NOT EXISTS (SELECT 1 FROM policies p WHERE p.client_id = c.id) " +
  "AND EXISTS (SELECT 1 FROM client_relations r WHERE r.related_client_id = c.id)";

function buildConditions(filter: ClientFilter): Conditions {
  const c = new Conditions();

  if (!filter.includeArchived) c.addRaw("c.is_archived = 0");

  const search = filter.search?.trim();

  // Searching reaches everybody. Someone typing a child's name is looking for
  // that child, and a book that held them but would not admit it when asked by
  // name would be worse than one that never held them at all.
  if (!filter.includeFamily && !search) c.addRaw(`NOT (${IS_DEPENDENT})`);

  if (search) {
    const query = ftsQuery(search);
    if (query !== null) {
      c.add("c.id IN (SELECT rowid FROM clients_fts WHERE clients_fts MATCH ?)", query);
    } else {
      const pattern = likePattern(search);
      c.addMany("(c.full_name LIKE ? ESCAPE '\\' OR c.phone LIKE ? ESCAPE '\\')", [pattern, pattern]);
    }
  }

  const city = blankToNull(filter.city);
  if (city) c.add("c.city = ?", city);
  const state = blankToNull(filter.state);
  if (state) c.add("c.state = ?", state);
  const category = blankToNull(filter.category);
  if (category) {
    c.add("EXISTS (SELECT 1 FROM policies p WHERE p.client_id = c.id AND p.category = ?)", category);
  }
  if (filter.missingEmail) c.addRaw("(c.email IS NULL OR c.email = '')");

  return c;
}

export function list(conn: Conn, filter: ClientFilter): Page<Client> {
  const conditions = buildConditions(filter);
  const whereSql = conditions.whereSql();

  const total = count(conn, `SELECT COUNT(*) AS n FROM clients c${whereSql}`, conditions.params());

  const { page, pageSize, limit, offset } = paginate(filter.page, filter.pageSize);
  const order = orderBy(filter.sort, filter.descending ?? false, SORTABLE, "c.full_name");

  const rows = conn
    .prepare(`SELECT ${COLUMNS}, ${DERIVED} FROM clients c${whereSql}${order} LIMIT ? OFFSET ?`)
    .all(...conditions.paramsWith([limit, offset])) as Record<string, unknown>[];

  return { rows: toModels<Client>(rows), total, page, pageSize };
}

export function get(conn: Conn, id: number): Client {
  const row = conn
    .prepare(`SELECT ${COLUMNS}, ${DERIVED} FROM clients c WHERE c.id = ?`)
    .get(id) as Record<string, unknown> | undefined;
  if (!row) throw AppError.notFound("Client");
  return toModel<Client>(row);
}

export function nextClientCode(conn: Conn): string {
  const row = conn
    .prepare(
      "SELECT IFNULL(MAX(CAST(substr(client_code, 4) AS INTEGER)), 0) + 1 AS next " +
        "FROM clients WHERE client_code GLOB 'CL-[0-9]*'",
    )
    .get() as { next: number };
  return `CL-${`${row.next}`.padStart(5, "0")}`;
}

function validate(input: ClientInput): void {
  if (input.fullName.trim() === "") throw AppError.validation("Client name is required");

  const email = blankToNull(input.email);
  if (email !== null && !looksLikeEmail(email)) {
    throw AppError.validation(`"${email}" is not a valid email address`);
  }

  const dob = blankToNull(input.dateOfBirth);
  if (dob !== null && parseDate(dob) === null) {
    throw AppError.validation("Date of birth is not a valid date");
  }
}

/** The column values shared by create and update, in the order both statements bind them. */
function fields(input: ClientInput): Bind[] {
  const dob = blankToNull(input.dateOfBirth);
  const pan = blankToNull(input.pan);
  const gstin = blankToNull(input.gstin);
  return [
    tidyName(input.fullName),
    blankToNull(input.email),
    input.phone ? normalisePhone(input.phone) : null,
    input.altPhone ? normalisePhone(input.altPhone) : null,
    dob === null ? null : parseDate(dob),
    blankToNull(input.gender),
    blankToNull(input.addressLine1),
    blankToNull(input.addressLine2),
    blankToNull(input.city),
    blankToNull(input.state),
    blankToNull(input.pincode),
    blankToNull(input.occupation),
    pan === null ? null : pan.toUpperCase(),
    gstin === null ? null : gstin.toUpperCase(),
    input.preferredLanguage ?? "en",
    boolToInt(input.remindersOptedOut),
    blankToNull(input.notes),
  ];
}

export function create(conn: Conn, input: ClientInput): number {
  validate(input);
  const code = blankToNull(input.clientCode) ?? nextClientCode(conn);

  try {
    const result = conn
      .prepare(
        "INSERT INTO clients (client_code, full_name, email, phone, alt_phone, date_of_birth, " +
          "gender, address_line1, address_line2, city, state, pincode, occupation, pan, gstin, " +
          "preferred_language, reminders_opted_out, notes) " +
          "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
      )
      .run(code, ...fields(input));
    return Number(result.lastInsertRowid);
  } catch (error) {
    throw mapUniqueError(error);
  }
}

export function update(conn: Conn, id: number, input: ClientInput): void {
  validate(input);

  let changes: number;
  try {
    const result = conn
      .prepare(
        "UPDATE clients SET full_name = ?, email = ?, phone = ?, alt_phone = ?, " +
          "date_of_birth = ?, gender = ?, address_line1 = ?, address_line2 = ?, " +
          "city = ?, state = ?, pincode = ?, occupation = ?, pan = ?, gstin = ?, " +
          "preferred_language = ?, reminders_opted_out = ?, notes = ?, " +
          "client_code = COALESCE(?, client_code) WHERE id = ?",
      )
      .run(...fields(input), blankToNull(input.clientCode), id);
    changes = result.changes;
  } catch (error) {
    throw mapUniqueError(error);
  }

  if (changes === 0) throw AppError.notFound("Client");
}

export function setArchived(conn: Conn, id: number, archived: boolean): void {
  const result = conn.prepare("UPDATE clients SET is_archived = ? WHERE id = ?").run(boolToInt(archived), id);
  if (result.changes === 0) throw AppError.notFound("Client");
}

/**
 * Archives or restores a client together with the people directly related to
 * them, and answers with how many rows it moved.
 *
 * One step out, deliberately. A family has no boundary of its own — it is
 * whoever the edges reach — so an operation that walked the whole graph would
 * grow every time an in-law was recorded, and putting one household away would
 * eventually put away half the book.
 */
export function setFamilyArchived(conn: Conn, id: number, archived: boolean): number {
  get(conn, id);
  const ids = [...immediateIds(conn, id), id];

  let moved = 0;
  for (const person of ids) {
    moved += conn
      .prepare("UPDATE clients SET is_archived = ? WHERE id = ? AND is_archived <> ?")
      .run(boolToInt(archived), person, boolToInt(archived)).changes;
  }
  return moved;
}

/**
 * Removes the client together with their policies and documents. Relationship
 * edges go with them, but the people on the other end do not: they are clients,
 * and one of them holding cover of their own is the ordinary case.
 *
 * The UI confirms first, naming who is involved; the archive flag is the softer
 * option offered alongside it.
 */
export function remove(conn: Conn, id: number): void {
  const result = conn.prepare("DELETE FROM clients WHERE id = ?").run(id);
  if (result.changes === 0) throw AppError.notFound("Client");
}

/**
 * Removes the client and the people directly related to them, answering with
 * every id it deleted so the interface can say what went.
 *
 * Reaches one step out for the same reason `setFamilyArchived` does. What it
 * deletes is therefore the family as it stood when the operator was shown the
 * list, not whatever the graph grows into later.
 */
export function removeWithImmediateFamily(conn: Conn, id: number): number[] {
  get(conn, id);
  const ids = [...immediateIds(conn, id), id];

  const deleted: number[] = [];
  for (const person of ids) {
    if (conn.prepare("DELETE FROM clients WHERE id = ?").run(person).changes > 0) {
      deleted.push(person);
    }
  }
  return deleted;
}

export function distinctCities(conn: Conn): string[] {
  const rows = conn
    .prepare("SELECT DISTINCT city FROM clients WHERE city IS NOT NULL AND city <> '' ORDER BY city")
    .all() as { city: string }[];
  return rows.map((row) => row.city);
}

/**
 * Finds an existing client by code, then email, then phone, then name — used by
 * the importer to decide between updating and inserting.
 */
export function findMatch(
  conn: Conn,
  code: string | null,
  email: string | null,
  phone: string | null,
  name: string,
): number | null {
  const attempts: [string, string | null][] = [
    ["SELECT id FROM clients WHERE client_code = ?", code],
    ["SELECT id FROM clients WHERE lower(email) = lower(?)", email],
    ["SELECT id FROM clients WHERE phone = ?", phone],
  ];

  for (const [sql, value] of attempts) {
    if (!value) continue;
    const row = conn.prepare(sql).get(value) as { id: number } | undefined;
    if (row) return row.id;
  }

  const row = conn.prepare("SELECT id FROM clients WHERE lower(full_name) = lower(?)").get(name) as
    | { id: number }
    | undefined;
  return row?.id ?? null;
}

function mapUniqueError(error: unknown): AppError {
  if (isConstraintViolation(error) && describe(error).includes("client_code")) {
    return AppError.conflict("That client code is already in use");
  }
  return AppError.database(error);
}
